### PR TITLE
Retry creating a subscription on gateway timeouts

### LINKED ISSUE #26: Retry subscriptions on receiving 504 response
We should retry (up to 5 times?) subscribing after a few seconds if the server responds with a gateway timeout (504).

### PR DESCRIPTION
**Pull request by joelwilliamsmith**
If the server responds to the subscribe request with a 504 (gateway
timeout) then we need to retry creating the subscription. Retries up to 5
times every 1 second then gives up and goes to error state.

Fixes #26

### CODE REVIEW

File: src/openapi/streaming/subscription.js [hunk 1]
```diff
@@ -218,13 +226,20 @@ function onSubscribeError(referenceId, response) {
...
-    this.currentState = STATE_UNSUBSCRIBED;
-    log.error(LOG_AREA, 'An error occurred subscribing', { response, url: this.url });
+    if (response.status === '504' && this.retryCounter++ < MAX_RETRIES) {
+        log.error(LOG_AREA, 'Received gateway timeout when subscribing. Retrying.', { response, url: this.url });
 
-    // if we are unsubscribed, do not fire the error handler
-    if (this.queue.peek() !== ACTION_UNSUBSCRIBE) {
-        if (this.onError) {
-            this.onError(response);
+        this.currentState = STATE_SUBSCRIBED; // assume we have an active (but faulty) subscription, so unsubscribe and try again
```
Comment (lukeapage) [inline src/openapi/streaming/subscription.js]: Comment this line that this is so unsubscribe isn't ignored?

File: src/openapi/streaming/subscription.js [hunk 2]
```diff
@@ -208,6 +211,11 @@ function onSubscribeSuccess(referenceId, result) {
     onReadyToPerformNextAction.call(this);
 }
 
+function retrySubscription() {
+    this.onUnsubscribe();
+    setTimeout(this.onSubscribe.bind(this), RETRY_DELAY);
```
Comment (lukeapage) [inline src/openapi/streaming/subscription.js]: Further checks in set timeout? Ex.g. What if we unsubscribed during timeout?

File: src/openapi/streaming/subscription.js [hunk 3]
```diff
@@ -295,6 +310,11 @@ function Subscription(streamingContextId, transport, serviceGroup, url, subscrip
     this.referenceId = null;
 
     /**
+     * The number of retries this subscription has made
+     */
+    this.retryCounter = 0;
```
Comment (lukeapage) [inline src/openapi/streaming/subscription.js]: Reset counter? Or I missed that?

File: src/openapi/streaming/subscription.js [hunk 4]
```diff
@@ -218,13 +226,20 @@ function onSubscribeError(referenceId, response) {
         return;
     }
 
-    this.currentState = STATE_UNSUBSCRIBED;
-    log.error(LOG_AREA, 'An error occurred subscribing', { response, url: this.url });
+    if (response.status === '504' && this.retryCounter++ < MAX_RETRIES) {
```
Comment (lukeapage) [inline src/openapi/streaming/subscription.js]: Do we previously assume response is an object or non null?

File: src/openapi/streaming/subscription.js (hunk 3 above)
Comment (joelwilliamsmith) [inline src/openapi/streaming/subscription.js]: Resetting in `onSubscribeSuccess`

**Comment by michax** [comment]
Suggestion to consider. Wouldn't it make more sense to increase delay? For example, start with 1000 ms delay, and end with 5000 ms for last retry.  We could use very natural Fibonacci sequence and use 1s, 1s, 2s, 3s, 5s. Giving us total of 12s for getting back on track. What do you think?

File: src/openapi/streaming/subscription.js (hunk 3 above)
Comment (michax) [inline src/openapi/streaming/subscription.js]: You should also reset it in subscription.reset method, and clear ongoing retry timeout if present.

I think we should also dispose of retries whenever new subscription request comes in. Especially when user requests unsubscribe.

File: src/openapi/streaming/subscription.js [hunk 5]
```diff
@@ -218,13 +228,20 @@ function onSubscribeError(referenceId, response) {
...
-    this.currentState = STATE_UNSUBSCRIBED;
-    log.error(LOG_AREA, 'An error occurred subscribing', { response, url: this.url });
+    if (response && response.status === '504' && this.retryCounter++ < MAX_RETRIES) {
+        log.error(LOG_AREA, 'Received gateway timeout when subscribing. Retrying.', { response, url: this.url });
 
-    // if we are unsubscribed, do not fire the error handler
-    if (this.queue.peek() !== ACTION_UNSUBSCRIBE) {
-        if (this.onError) {
-            this.onError(response);
+        this.currentState = STATE_SUBSCRIBED; // set to subscribed so that unsubscribe request isn't ignored
```
Comment (michax) [inline src/openapi/streaming/subscription.js]: Ideally I think we should peek next action, and check if it's unsubscribe. If so, no need for retrying.

File: src/openapi/streaming/subscription.js [hunk 6]
```diff
@@ -218,13 +228,20 @@ function onSubscribeError(referenceId, response) {
...
+        retrySubscription.call(this);
```
Comment (michax) [inline src/openapi/streaming/subscription.js]: It looks like we will run two actions in such case. First, retrySubscription will run unsubscribe immediately. And at the end of this function we are calling:  
```
onReadyToPerformNextAction.call(this);
```
Which will run next action from a queue. I don't think it's quite right.

Comment (michax) [inline src/openapi/streaming/subscription.js]: It looks like we will run two actions in such case. First, retrySubscription will run unsubscribe immediately. And at the end of this function we are calling:  
```
onReadyToPerformNextAction.call(this);
```
Which will run next action from a queue. I don't think it's quite right.

File: test/unit/openapi/streaming/subscription-spec.js [hunk 7]
```diff
@@ -112,6 +112,52 @@ describe("openapi StreamingSubscription", () => {
...
+                    tick(() => {
+                        expect(errorSpy).not.toHaveBeenCalled();
+                        expect(createdSpy.calls.count()).toEqual(1);
+                        done();
+                    });
+                });
+            });
+        });
+
+        it("retries subscribing 5 times before giving up", (done) => {
```
Comment (michax) [inline test/unit/openapi/streaming/subscription-spec.js]: I think we need more tests for edge cases. 
* Unsubscribe request in the middle of retry.
* Modify request in the middle of retry.
* Subscribe request in the middle of retry. 
* Dispose (reset) request in the middle of retry.
* Modify which has timing out unsubscribe. But needs to follow up with subscribe afterwards (from queue).

File: src/openapi/streaming/subscription.js [hunk 8]
```diff
@@ -205,9 +208,16 @@ function onSubscribeSuccess(referenceId, result) {
     }
     this.updatesBeforeSubscribed = null;
 
+    this.retryCounter = 0;
+
     onReadyToPerformNextAction.call(this);
 }
 
+function retrySubscription() {
+    this.onUnsubscribe();
```
Comment (michax) [inline src/openapi/streaming/subscription.js]: If Gateway timetous on subscribe request. Why do we need to unsubscribe? 
Also, what will happend if onUnsubscribeError returns timeout as well?

File: src/openapi/streaming/subscription.js (hunk 5 above)
Comment (michax) [inline src/openapi/streaming/subscription.js]: What do you think about maybe including STATE_SUBSCRIBE_TIMEOUT? Seems a little confusing to override this.currentState with non factual one.

File: src/openapi/streaming/subscription.js (hunk 6 above)
Comment (michax) [inline src/openapi/streaming/subscription.js]: If we indeed need to call unsubscribe first to subscribe after subscribe timeout. We should first peek into next action. If next action is unsubscribe, no need to try to retry subscription. 
```
this.queue.peek()
```

File: src/openapi/streaming/subscription.js (hunk 8 above)
Comment (joelwilliamsmith) [inline src/openapi/streaming/subscription.js]: It could be that a proxy timed out, and the subscription was actually successful. I _assume_ the server copes with receiving an unsubscribe request for a non-existent subscription if that's not indeed the case.

**Comment by lukeapage** [comment]
superseded by #29
